frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+		#[pallet::call_index(0)]
+		#[pallet::weight(
+			T::WeightInfo::initialize_pallet_grow(witness_count.unwrap_or_default())
+				.max(T::WeightInfo::initialize_pallet_shrink(witness_count.unwrap_or_default()))
+		)]
+		pub fn initialize_pallet(
+			origin: OriginFor<T>,
+			new_count: u32,
+			witness_count: Option<u32>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let current_count = TrashDataCount::<T>::get();
+			ensure!(
+				current_count == witness_count.unwrap_or_default(),
+				Error::<T>::AlreadyInitialized
+			);
+
+			// Add or remove elements from `TrashData`.
+			(new_count..current_count).for_each(|i| TrashData::<T>::remove(i));
+			(current_count..new_count).for_each(|i| TrashData::<T>::insert(i, [i as u8; 1024]));
+
+			Self::deposit_event(Event::PalletInitialized { reinit: witness_count.is_some() });
+			TrashDataCount::<T>::set(new_count);
+			Ok(())
+		}
+
+		/// Set the `Compute` storage value that determines how much of the
+		/// block's weight `ref_time` to use during `on_idle`.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::set_compute())]
+		pub fn set_compute(origin: OriginFor<T>, compute: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Compute::<T>::set(compute);
+
+			Self::deposit_event(Event::ComputationLimitSet { compute });
+			Ok(())
+		}
+
+		/// Set the `Storage` storage value that determines the PoV size usage
+		/// for each block.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::set_storage())]
+		pub fn set_storage(origin: OriginFor<T>, storage: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Storage::<T>::set(storage);
+
+			Self::deposit_event(Event::StorageLimitSet { storage });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Waste at most the remaining proof size of `meter`.
+		///
+		/// Tries to come as close to the limit as possible.
+		pub(crate) fn waste_at_most_proof_size(meter: &mut WeightMeter) {
+			let Ok(n) = Self::calculate_proof_size_iters(&meter) else {
+				return;
+			};
+
+			meter.defensive_saturating_accrue(T::WeightInfo::waste_proof_size_some(n));
+
+			(0..n).for_each(|i| {
+				TrashData::<T>::get(i);
+			});
+		}
+
+		/// Calculate how many times `waste_proof_size_some` should be called to fill up `meter`.
+		fn calculate_proof_size_iters(meter: &WeightMeter) -> Result<u32, ()> {
+			let base = T::WeightInfo::waste_proof_size_some(0);
+			let slope = T::WeightInfo::waste_proof_size_some(1).saturating_sub(base);
+
+			let remaining = meter.remaining().saturating_sub(base);
+			let iter_by_proof_size =
+				remaining.proof_size().checked_div(slope.proof_size()).ok_or(())?;
+			let iter_by_ref_time = remaining.ref_time().checked_div(slope.ref_time()).ok_or(())?;
+
+			match iter_by_proof_size {
+				0 => Err(()),
+				i if i <= iter_by_ref_time => Ok(i as u32),
+				_ => Err(()),
+			}
```

Comment:
Stylistic: this could be reduced into an `if`:

```suggestion
			if iter_by_proof_size > 0 && iter_by_proof_size <= iter_by_ref_time {
				Ok(iter_by_proof_size as u32)
			} else {
				Err(())
			}
```